### PR TITLE
Add support for effectful conversions

### PR DESCRIPTION
**Pull request by jcp19**
Adds support for conversions which might produce side effects (e.g. converting from `string` to `[]byte`). Furthermore, this fixes an unsoundness caused by issue #246.

### CODE REVIEW

File: src/main/scala/viper/gobra/translator/encodings/StringEncoding.scala [hunk 1]
```diff
@@ -68,13 +75,61 @@ class StringEncoding extends LeafTypeEncoding {
...
+        // interesting properties
+        val (pos, info, errT) = conv.vprMeta
+
+        val sliceT = in.SliceT(in.IntT(Addressability.sliceElement, TypeBounds.Byte), Addressability.outParameter)
+        val slice = in.LocalVar(Names.freshName, sliceT)(conv.info)
+        val vprSlice = ctx.typeEncoding.variable(ctx)(slice)
+        val qtfVar = in.BoundVar("i", in.IntT(Addressability.boundVariable))(conv.info)
+        val post = in.SepForall(
+          vars = Vector(qtfVar),
+          triggers = Vector(),
```
Comment (ArquintL) [inline src/main/scala/viper/gobra/translator/encodings/StringEncoding.scala:110]: could you use the indexed expression as a trigger?

File: src/main/scala/viper/gobra/translator/encodings/StringEncoding.scala [hunk 2]
```diff
@@ -68,13 +75,61 @@ class StringEncoding extends LeafTypeEncoding {
...
+    }
+  }
+
+
+  /**
+    * Encodes the (effectul) conversion from a string to a []byte
+    * [ target = []byte(str) ] ->
+    *   [
+    *     var s []byte
+    *     inhale forall i Int :: 0 <= i && i < len(s) ==> acc(&s[i])
```
Comment (ArquintL) [inline src/main/scala/viper/gobra/translator/encodings/StringEncoding.scala:90]: is it correct that you (currently) only take care of the permissions without restricting the resulting values in any way?

File: src/main/scala/viper/gobra/translator/encodings/StringEncoding.scala [hunk 3]
```diff
@@ -209,4 +264,41 @@ class StringEncoding extends LeafTypeEncoding {
...
+  private val byteSliceToStrFuncGenerator: FunctionGenerator[Unit] = new FunctionGenerator[Unit] {
+    override def genFunction(@unused x: Unit)(ctx: Context): vpr.Function = {
+      val info = Source.Parser.Internal
+      val paramT = in.SliceT(in.IntT(Addressability.sliceElement, TypeBounds.Byte), Addressability.outParameter)
+      val param = in.Parameter.In("s", paramT)(info)
+      val res = in.Parameter.Out("res", in.StringT(Addressability.outParameter))(info)
+      val qtfVar = in.BoundVar("i", in.IntT(Addressability.boundVariable))(info)
+      val pre = in.SepForall(
+        vars = Vector(qtfVar),
+        triggers = Vector(),
```
Comment (ArquintL) [inline src/main/scala/viper/gobra/translator/encodings/StringEncoding.scala:282]: same here

**Review by ArquintL — APPROVED** [review]
a few smaller remarks

File: src/main/scala/viper/gobra/translator/encodings/StringEncoding.scala (hunk 2 above)
Comment (jcp19) [reply src/main/scala/viper/gobra/translator/encodings/StringEncoding.scala:90]: Yes, that is the point. We can enrich the specification once we introduce more string operations (e.g. indexing strings)

**Pull request merged**